File: relay/test/frames.test.ts
import { describe, it, expect } from "vitest";
import { FrameType, encodeFrame, decodeFrame } from "../src/frames.js";
import { encodeJsonFrame, decodeJsonPayload } from "../src/frames.js";

describe("FrameType enum", () => {
  it("has the frozen type codes", () => {
    expect(FrameType.AUTH).toBe(0x01);
    expect(FrameType.AUTH_OK).toBe(0x02);
    expect(FrameType.AUTH_FAIL).toBe(0x03);
    expect(FrameType.OPEN).toBe(0x10);
    expect(FrameType.OPEN_OK).toBe(0x11);
    expect(FrameType.OPEN_FAIL).toBe(0x12);
    expect(FrameType.DATA).toBe(0x20);
    expect(FrameType.CLOSE).toBe(0x21);
    expect(FrameType.PING).toBe(0x30);
    expect(FrameType.PONG).toBe(0x31);
  });
});

describe("encodeFrame / decodeFrame", () => {
  it("round-trips an empty-payload control frame", () => {
    const buf = encodeFrame(FrameType.AUTH_OK, 0, Buffer.alloc(0));
    expect(buf.length).toBe(5);
    expect(buf[0]).toBe(0x02);
    const f = decodeFrame(buf);
    expect(f.type).toBe(FrameType.AUTH_OK);
    expect(f.streamId).toBe(0);
    expect(f.payload.length).toBe(0);
  });
});

describe("encodeFrame round-trips every frame type", () => {
  const cases: Array<[FrameType, number, Buffer]> = [
    [FrameType.AUTH, 0, Buffer.from('{"pairingToken":"t"}', "utf8")],
    [FrameType.AUTH_OK, 0, Buffer.alloc(0)],
    [FrameType.AUTH_FAIL, 0, Buffer.from('{"reason":"bad"}', "utf8")],
    [FrameType.OPEN, 7, Buffer.from('{"host":"a","port":443}', "utf8")],
    [FrameType.OPEN_OK, 7, Buffer.alloc(0)],
    [FrameType.OPEN_FAIL, 7, Buffer.from('{"reason":"refused"}', "utf8")],
    [FrameType.DATA, 9, Buffer.from([1, 2, 3, 4, 5])],
    [FrameType.CLOSE, 9, Buffer.alloc(0)],
    [FrameType.PING, 0, Buffer.alloc(0)],
    [FrameType.PONG, 0, Buffer.alloc(0)],
  ];

  for (const [type, streamId, payload] of cases) {
    it(`round-trips type 0x${type.toString(16)} stream ${streamId}`, () => {
      const enc = encodeFrame(type, streamId, payload);
      const dec = decodeFrame(enc);
      expect(dec.type).toBe(type);
      expect(dec.streamId).toBe(streamId);
      expect(Buffer.compare(dec.payload, payload)).toBe(0);
    });
  }

  it("preserves a large 32-bit streamId via big-endian", () => {
    const enc = encodeFrame(FrameType.DATA, 0xdeadbeef, Buffer.from([0xff]));
    expect(enc[1]).toBe(0xde);
    expect(enc[2]).toBe(0xad);
    expect(enc[3]).toBe(0xbe);
    expect(enc[4]).toBe(0xef);
    expect(decodeFrame(enc).streamId).toBe(0xdeadbeef);
  });

  it("throws on a buffer shorter than the header", () => {
    expect(() => decodeFrame(Buffer.from([0x20, 0x00]))).toThrow();
  });
});

describe("JSON frame helpers", () => {
  it("round-trips a JSON object", () => {
    const enc = encodeJsonFrame(FrameType.OPEN, 42, { host: "x.com", port: 80 });
    const dec = decodeFrame(enc);
    expect(dec.type).toBe(FrameType.OPEN);
    expect(dec.streamId).toBe(42);
    expect(decodeJsonPayload(dec.payload)).toEqual({ host: "x.com", port: 80 });
  });
});

File: android/app/src/test/java/com/tetherproxy/app/tunnel/FramesTest.kt
package com.tetherproxy.app.tunnel

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class FramesTest {

    @Test
    fun typeCodes_matchFrozenProtocol() {
        assertEquals(0x01, FrameType.AUTH.code)
        assertEquals(0x02, FrameType.AUTH_OK.code)
        assertEquals(0x03, FrameType.AUTH_FAIL.code)
        assertEquals(0x10, FrameType.OPEN.code)
        assertEquals(0x11, FrameType.OPEN_OK.code)
        assertEquals(0x12, FrameType.OPEN_FAIL.code)
        assertEquals(0x20, FrameType.DATA.code)
        assertEquals(0x21, FrameType.CLOSE.code)
        assertEquals(0x30, FrameType.PING.code)
        assertEquals(0x31, FrameType.PONG.code)
    }

    @Test
    fun fromCode_resolvesEveryType() {
        for (t in FrameType.values()) {
            assertEquals(t, FrameType.fromCode(t.code))
        }
        assertNull(FrameType.fromCode(0x99))
    }

    @Test
    fun encode_layoutIsTypeThenBigEndianStreamIdThenPayload() {
        val bytes = Frames.encode(FrameType.DATA, 0x01020304, byteArrayOf(0xAA.toByte(), 0xBB.toByte()))
        assertArrayEquals(
            byteArrayOf(0x20, 0x01, 0x02, 0x03, 0x04, 0xAA.toByte(), 0xBB.toByte()),
            bytes
        )
    }

    @Test
    fun encode_streamIdZeroForControlFrames() {
        val bytes = Frames.encode(FrameType.PING, 0, ByteArray(0))
        assertArrayEquals(byteArrayOf(0x30, 0x00, 0x00, 0x00, 0x00), bytes)
    }

    @Test
    fun encode_handlesHighBitStreamId() {
        val bytes = Frames.encode(FrameType.CLOSE, 0xFFFFFFFF.toInt(), ByteArray(0))
        assertArrayEquals(
            byteArrayOf(0x21, 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte(), 0xFF.toByte()),
            bytes
        )
    }

    @Test
    fun decode_roundTripsEveryFrameType() {
        val payloads = mapOf(
            FrameType.AUTH to """{"pairingToken":"t"}""".toByteArray(),
            FrameType.AUTH_OK to ByteArray(0),
            FrameType.AUTH_FAIL to """{"reason":"bad"}""".toByteArray(),
            FrameType.OPEN to """{"host":"api.example.com","port":443}""".toByteArray(),
            FrameType.OPEN_OK to ByteArray(0),
            FrameType.OPEN_FAIL to """{"reason":"refused"}""".toByteArray(),
            FrameType.DATA to byteArrayOf(1, 2, 3, 4, 5),
            FrameType.CLOSE to ByteArray(0),
            FrameType.PING to ByteArray(0),
            FrameType.PONG to ByteArray(0)
        )
        var streamId = 7
        for ((type, payload) in payloads) {
            val encoded = Frames.encode(type, streamId, payload)
            val decoded = Frames.decode(encoded)
            assertEquals(type, decoded.type)
            assertEquals(streamId, decoded.streamId)
            assertArrayEquals(payload, decoded.payload)
            streamId++
        }
    }

    @Test
    fun decode_unknownTypeThrows() {
        try {
            Frames.decode(byteArrayOf(0x99.toByte(), 0, 0, 0, 1))
            throw AssertionError("expected IllegalArgumentException")
        } catch (e: IllegalArgumentException) {
            // expected
        }
    }

    @Test
    fun decode_tooShortThrows() {
        try {
            Frames.decode(byteArrayOf(0x20, 0, 0))
            throw AssertionError("expected IllegalArgumentException")
        } catch (e: IllegalArgumentException) {
            // expected
        }
    }

    @Test
    fun jsonHelpers_buildAndParseControlPayloads() {
        val auth = Frames.authJson(
            pairingToken = "tok",
            deviceId = "dev-1",
            proxyUsername = "user",
            proxyPassword = "pass"
        )
        val parsed = Frames.parseJson(auth)
        assertEquals("tok", parsed["pairingToken"])
        assertEquals("dev-1", parsed["deviceId"])
        assertEquals("user", parsed["proxyUsername"])
        assertEquals("pass", parsed["proxyPassword"])

        val open = """{"host":"h.example.com","port":8443}"""
        val openParsed = Frames.parseJson(open)
        assertEquals("h.example.com", openParsed["host"])
        assertEquals("8443", openParsed["port"])
    }
}
